src/Microsoft.AspNetCore.Blazor.Server/MonoDebugProxy/MonoDebugProxyAppBuilderExtensions.cs

```diff
@@ -0,0 +1,178 @@
+            {
+                await context.Response.WriteAsync($@"
+                    <h1>Unable to find debuggable browser tab</h1>
+                    <p>
+                        The response from <code>{debuggerTabsListUrl}</code> does not include
+                        any entry for <code>{targetTabUrl}</code>.
+                    </p>");
+                return;
+            }
+            else if (matchingTabs.Count > 1)
+            {
+                // TODO: Automatically disambiguate by adding a GUID to the page title
+                // when you press the debugger hotkey, include it in the querystring passed
+                // here, then remove it once the debugger connects.
+                await context.Response.WriteAsync($@"
+                    <h1>Multiple matching tabs are open</h1>
+                    <p>
+                        There is more than one browser tab at <code>{targetTabUrl}</code>.
+                        Close the ones you do not wish to debug, then refresh this page.
+                    </p>");
+                return;
+            }
+
+            // Now we know uniquely which tab to debug, construct the URL to the debug
+            // page and redirect there
+            var tabToDebug = matchingTabs.Single();
+            var underlyingV8Endpoint = tabToDebug.WebSocketDebuggerUrl;
+            var proxyEndpoint = $"{request.Host}{request.PathBase}/_framework/debug/ws-proxy?browser={WebUtility.UrlEncode(underlyingV8Endpoint)}";
+            var devToolsUrlAbsolute = new Uri(debuggerHost + tabToDebug.DevtoolsFrontendUrl);
+            var devToolsUrlWithProxy = $"{devToolsUrlAbsolute.Scheme}://{devToolsUrlAbsolute.Authority}{devToolsUrlAbsolute.AbsolutePath}?ws={proxyEndpoint}";
+            context.Response.Redirect(devToolsUrlWithProxy);
+        }
+
+        private static string GetLaunchChromeInstructions(string appRootUrl)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return $@"<p>Close this browser, then press Win+R and enter the following:</p>
+                          <p><strong><code>""%programfiles(x86)%\Google\Chrome\Application\chrome.exe"" --remote-debugging-port=9222 {appRootUrl}</code></strong></p>";
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                return $@"<p>Close this browser, then in a terminal window execute the following:</p>
+                          <p><strong><code>google-chrome --remote-debugging-port=9222 {appRootUrl}</code></strong></p>";
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                return $@"<p>Close this browser, then in a terminal window execute the following:</p>
+                          <p><strong><code>open /Applications/Google\ Chrome.app --remote-debugging-port=9222 {appRootUrl}</code></strong></p>";
+            }
```

Comment:
Is this the default or will users have to launch chrome this way every time? (Does VS do this for instance)?

Reply:
Unfortunately it's not the default. You can change your own VS config to launch Chrome with `--remote-debugging-port=9222` by editing the list of browsers in the "Browse With" dialog, but I'm not aware of a way of putting that config in `launchSettings.json`.

Long term we might want to ask about launching Chrome with that flag by default for Blazor project types.

For now people will just have to follow the instructions to open a new instance. Fortunately it's only once: for as long as that browser instance remains open, VS will open new tabs in it each time you Ctrl+F5.